### PR TITLE
📖 Kubernetes support matrix

### LINKED ISSUE #1518: [docs] Kubernetes support matrix
/kind documentation

Write down supported Kubernetes versions. The questions I would like answered are:

1. What versions of kubernetes can I install with a given version of Cluster API?
2. Do we ever drop support for a kubernetes version?

To understand why this is a difficult question let's examine Cluster API v1alpha2. When we started it, CAPI v1alpha1 supported Kubernetes v1.13, v1.14 and v1.15. By the time work on v1alpha3 began, v1alpha2 supported v1.13, 1.14, 1.15 and 1.16. Do we stick with an n-2 versioning scheme? Do we do something more complex like: We support all versions spanning (n-2, n-1, n (== current version)) as well as all releases released during the CAPI development cycle? For example, CAPI v1a2 supports v1.13-v1.16 because those were the valid k8s versions during development. This is just an idea. We could certainly drop support as k8s rolls out new versions, but then we get into a weird place where we drop support in the middle of a CAPI development cycle. Realistically though that should be fine.

### PR DESCRIPTION
**Pull request by dlipovetsky**
**TODO**
- [ ] links to glossary from project jargon (e.g. workload, management cluster)
- [x] clusterctl support (separate issue, https://github.com/kubernetes-sigs/cluster-api/issues/3090)
- [x] Upgrade order (separate issue, https://github.com/kubernetes-sigs/cluster-api/issues/3091)

**What this PR does / why we need it**:
Documents supported versions and and compatible Kubernetes API versions

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1518

### CODE REVIEW

File: docs/book/src/reference/version-skew-policy.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Cluster API Version and version skew support policy
+
+## Supported versions
+
+The Cluster API project maintains the Core Provider, the Kubeadm Bootstrap Provider, the Kubeadm Control Plane Provider. These run on a Management Cluster. The project aims to keep the current minor release compatible with the actively supported Kubernetes minor releases, i.e., the current release (N), N-1, and N-2. To find out the exact range of Kubernetes APIs the Management Cluster may run, see the tables below.
+
+### Core Provider - `cluster-api-controller`
+
+||Cluster API v1alpha1 (v0.1)|Cluster API v1alpha2 (v0.2)|Cluster API v1alpha3 (v0.3)|
```
Comment (vincepri) [inline docs/book/src/reference/version-skew-policy.md:9]: Do we want to include v1alpha1? IIRC there is no easy upgrade path other from v1alpha1 -> v1alpha2, I'd remove it

File: docs/book/src/reference/version-skew-policy.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+|-|-|-|-|
+|Kubernetes v1.13||||
+|Kubernetes v1.14||✓||
+|Kubernetes v1.15||✓||
+|Kubernetes v1.16||✓|✓|
+|Kubernetes v1.17|||✓|
+|Kubernetes v1.18|||✓|
+
+The Kubeadm Bootstrap Provider generates configuration using the kubeadm APIs.
```
Comment (fabriziopandini) [inline docs/book/src/reference/version-skew-policy.md:33]: might be you want to add that we are using v1beta1 version of the kubeadm API

Comment (dlipovetsky) [reply docs/book/src/reference/version-skew-policy.md:39]: Will do, thanks!

I don't have a section on upgrading yet. Do you think there is anything CAPI users need to be aware of as v1beta1  is deprecated and eventually removed?

Comment (fabriziopandini) [reply docs/book/src/reference/version-skew-policy.md:39]: There is an issue for this, but no decision yet AFAIK

File: docs/book/src/reference/version-skew-policy.md [hunk 3]
```diff
@@ -1,60 +1,68 @@
...
 ### Infrastructure Providers
 
-Infrastructure Providers are _independently supported_. An Infrastructure Provider that is compatible with Cluster API version N must be compatible with at least one Kubernetes API version supported by Cluster API version N. An Infrastructure Provider may support a smaller set of versions than Cluster API version N.
+Infrastructure Providers are _independently maintained_. If an Infrastructure Provider version M is compatible with Cluster API version N, then M must be compatible with at least one Kubernetes API version supported by Cluster API version N. The Infrastructure Provider version M may be compatible with a smaller set of Kubernetes versions than Cluster API version N.
 
 For version compatibility, refer to each Infrastructure Provider's documentation. See this list of [public Infrastructure Providers](providers.md).
 
 ### Visual Representation
 
 #### Management And Workload Cluster Are the Same
```
Comment (fabriziopandini) [inline docs/book/src/reference/version-skew-policy.md:62]:
```suggestion
#### Management And Workload Cluster Are the Same (Self-hosted)
```

Comment (dlipovetsky) [reply docs/book/src/reference/version-skew-policy.md]: I'll change this, thanks

Comment (dlipovetsky) [reply docs/book/src/reference/version-skew-policy.md]: I saw your comments just after I pushed an update. I think it addresses the feedback you had for Infrastructure Providers.

**Comment by vincepri** [comment]
LGTM pending squash

**Comment by dlipovetsky** [comment]
Thanks for the reviews, @vincepri @CecileRobertMichon @fabriziopandini!

I'd like to cover clusterctl in this doc. And upgrades. I can do that in a separate PR, though.

**Comment by vincepri** [comment]
Got it, it probably makes sense to continue in this PR, wdyt?

**Comment by dlipovetsky** [comment]
> Got it, it probably makes sense to continue in this PR, wdyt?

SGTM. Since I saw the lgtm's, I wanted to make sure you were ok with reviewing additional changes :smile:

File: docs/book/src/reference/versions.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+All Infrastructure Providers are maintained by independent teams. Other Bootstrap and Control Plane Providers are also maintained by independent teams. For more information about their version support, see [below](#providers-maintained-by-independent-teams).
+
+## Supported Kubernetes Versions
+
+The project aims to keep the current minor release compatible with the actively supported Kubernetes minor releases, i.e., the current release (N), N-1, and N-2. To find out the exact range of Kubernetes versions supported by each  component, please see the [tables](#release-components) below.
+
+See the [following section](#kubernetes-version-support-as-a-function-of-cluster-toplogy) to understand how cluster topology affects version support.
+
+### Kubernetes Version Support As A Function Of Cluster Toplogy
```
Comment (sedefsavas) [inline docs/book/src/reference/versions.md:22]:
```suggestion
### Kubernetes Version Support As A Function Of Cluster Topology
```

Comment (dlipovetsky) [reply docs/book/src/reference/versions.md:22]: Thank you!

**Comment by vincepri** [comment]
/hold

@dlipovetsky squash please :)

**Comment by dlipovetsky** [comment]
Fixing typo and squashing. I wanted to cover clusterctl and upgrades, but haven't been able to. I'll revisit.

**Pull request merged**